tests/math/test_functions.py

```diff
@@ -1392,3 +1441,178 @@ def test_torch_coercion(self):
+        assert all(not isinstance(a, np.tensor) for a in unwrapped_params)
+
+    def test_autograd_unwrapping_backward(self):
+        """Test that a sequence of Autograd values is properly unwrapped
+        during the backward pass"""
+        unwrapped_params = None
+
+        def cost_fn(*params):
+            nonlocal unwrapped_params
+            unwrapped_params = qml.math.unwrap(params)
+            return np.sum(np.sin(params[0] * params[2])) + params[1]
+
+        values = [onp.array([0.1, 0.2]), np.tensor(0.1, dtype=np.float64), np.tensor([0.5, 0.2])]
+        grad = qml.grad(cost_fn)(*values)
+
+        expected = [np.array([0.1, 0.2]), 0.1, np.array([0.5, 0.2])]
+        assert all(np.allclose(a, b) for a, b in zip(unwrapped_params, expected))
+        assert all(not isinstance(a, ArrayBox) for a in unwrapped_params)
+
+    def test_autograd_unwrapping_backward_nested(self):
+        """Test that a sequence of Autograd values is properly unwrapped
+        during multiple backward passes"""
+        unwrapped_params = None
+
+        def cost_fn(p, max_depth=None):
+            nonlocal unwrapped_params
+            unwrapped_params = qml.math.unwrap(p, max_depth)
+            return np.sum(np.sin(np.prod(p)))
+
+        values = np.tensor([0.1, 0.2, 0.3])
+        hess = qml.jacobian(qml.grad(cost_fn))(values)
+
+        expected = np.array([0.1, 0.2, 0.3])
+        assert np.allclose(unwrapped_params, expected)
+        assert not isinstance(unwrapped_params, ArrayBox)
+
+        # Specifying max_depth=1 will result in the second backward
+        # pass not being unwrapped
+        hess = qml.jacobian(qml.grad(cost_fn))(values, max_depth=1)
+        assert all(isinstance(a, ArrayBox) for a in unwrapped_params)
+
+    def test_jax_unwrapping(self):
+        """Test that a sequence of Autograd values is properly unwrapped
+        during the forward pass"""
+        unwrapped_params = None
+
+        def cost_fn(params):
+            nonlocal unwrapped_params
+            unwrapped_params = qml.math.unwrap(params)
+            return np.sum(np.sin(params[0])) + params[2]
+
+        values = [jnp.array([0.1, 0.2]), onp.array(0.1, dtype=np.float64), jnp.array([0.5, 0.2])]
+        cost_fn(values)
+
+        expected = [np.array([0.1, 0.2]), 0.1, np.array([0.5, 0.2])]
+        assert all(np.allclose(a, b) for a, b in zip(unwrapped_params, expected))
+        assert all(not isinstance(a, np.tensor) for a in unwrapped_params)
+
+
+class TestGetTrainable:
+    """Test getting trainable indices"""
+
+    def test_tensorflow(self):
+        """Test that the trainability indices of a sequence of TensorFlow values
+        is correctly extracted"""
+        values = [
+            onp.array([0.1, 0.2]),
+            tf.Variable(0.1, dtype=tf.float64),
+            tf.constant([0.5, 0.2]),
+        ]
+
+        # outside of a gradient tape, no indices are trainable

```

Comment:
Wow, that's also clever, haven't thought about it

Reply:
It is new behaviour in TF 2.3+ 😆 In earlier versions, `Variable` objects were always trainable, and `constants` never trainable, regardless of a tape being present or not.